### PR TITLE
Replace chokidar with esbuild watch mode

### PR DESCRIPTION
**Pull request by Zertz**
This is very much a quickly hacked together proof of concept so the implementation isn't exactly nice 😅 

I removed chokidar and a couple other no deps that were no longer necessary and enabled esbuild's built-in watch mode. I tested the two examples and as far as I can tell, it seems to work.

I wasn't able however to get the webpack example to start as I was getting `TypeError: require(...) is not a function`. I didn't really look into it yet.

If it's a change you would consider merging, I'd gladly take your input on how to structure this.

### CODE REVIEW

**Review by Kiyozz — COMMENTED** [review]
(update package-lock) Why this commit at first?

**Comment by Zertz** [comment]
I had some trouble because we apparently don't run the same npm version and the lock format constantly changes and/or breaks 😞 

Either way, I'll clean it up along with the rest once we settle for an implementation!

**Comment by Kiyozz** [comment]
I'm working on it, will test this implementation

**Comment by Kiyozz** [comment]
Ok so, I made some changes.

But, the Esbuild API `build({ watch })` does not work for me. When I edit App.tsx, no event occurred: onRebuild is not fired.

I'm working on macOS. Did it works for you?

**Comment by Zertz** [comment]
I tested on macOS as well, did it work before your changes?

I can't test at the moment but I'm pretty sure calling `start()` in `onRebuild` isn't enough to trigger livereload?

**Comment by Kiyozz** [comment]
After reviewing my changes, I notice an error on onRebuild, I think I confused main and renderer.

I will do more tests today.

`start()` only destroy last electron and spawn a new one.

**Comment by Kiyozz** [comment]
I have successfully used esbuild.watch in main and renderer. There is however a problem -> we lose the fast-refresh of the css files.

Summary: chokidar has been removed to use livereload's watch function. However livereload uses chokidar internally. The question to ask is: is it really useful to remove chokidar in this case? One of the benefits is that we are removing the use of old/unmaintained library (dependency-tree is not up to date to the latest version of TypeScript, I had to fork it myself).

Is it a big deal to lose the css fast-refresh (esbuild in renderer)?

**Comment by Zertz** [comment]
You raise good points! I agree that there is little benefit if another dependency uses chokidar internally.

Someone [mentioned](https://github.com/evanw/esbuild/issues/21#issuecomment-770586792) using onRebuild with SSE (Server Sent Events) which seems like a pretty good idea. This should make it possible to also remove livereload. I'll put together a prototype!

File: src/builder/esbuild.ts [hunk 1]
```diff
@@ -34,110 +34,130 @@ export class EsbuildBuilder extends BaseBuilder<BuildOptions> {
...
-            }
-          })
 
-          const server = createServer(handler)
+      const handler = connect()
 
-          server.on('upgrade', (req, socket, head) => {
-            proxyLr.ws(req, socket, head)
-          })
+      // handler.use(compression() as never)
```
Comment (Zertz) [inline src/builder/esbuild.ts]: `compression` seems to break SSE. I'm assuming there's a fix but I haven't looked into it yet.

**Comment by Zertz** [comment]
Got it running with SSE! I'll test it on the Tailwind example next to see how it behaves with large amounts of CSS 🤞

**Comment by Kiyozz** [comment]
SSE looks promising. However, we still lose the fast-refresh of css. But this is due to esbuild.watch not notifying which file has been modified, and also because livereload has been deleted.

**Comment by Zertz** [comment]
I'll take a look into the CSS soon! Since everything is bundled into a single CSS and JS file, I'm thinking it should be possible to figure out what changed.

I also posted an [issue/question](https://github.com/evanw/esbuild/issues/1047) on esbuild regarding changed files.

File: src/builder/esbuild.ts (hunk 1 above)
Comment (Zertz) [reply src/builder/esbuild.ts]: https://github.com/expressjs/compression#server-sent-events

**Comment by Zertz** [comment]
I got CSS hot reload working in the `react-typescript` example! 🔥 

Still need to figure out why `onRebuild` isn't triggered in the Tailwind example 🤔 

I'm not sure in the end if this is all going to be worth it but I enjoy doing this exploration work nonetheless 👀

**Comment by Kiyozz** [comment]
For tailwindcss examples, this may be du to postcss plugin. With esbuild v0.10.0 plugins can now return a list of files to watch for changes. Below v0.10.0, esbuild was not aware of files changed by plugins.

**Comment by Kiyozz** [comment]
Need to rebase changes with latest main

**Comment by Zertz** [comment]
I had a feeling PostCSS was the issue, amazing timing on that 0.10.0 release 😄 

I'll rebase and keep an eye on updates for esbuild-plugin-postcss2 or perhaps send a PR there if time allows.

**Comment by Kiyozz** [comment]
The plugin doesn’t seem much maintained. There is already a PR to fix the peerDependencies, which hangs on a version of esbuild 0.8.x

I will say it is possible to fork it :)

**Comment by Kiyozz** [comment]
esbuild v0.10.1

CleanShot 2021-03-26 at 10 10 12

Yeah !

**Comment by Zertz** [comment]
Sweet! Hopefully that will make change detection easier 🤞

**Comment by Zertz** [comment]
`metafile` is neat but doesn't really help us here, comparing hashes still seems like the way to go

**Pull request closed without merging**